infrastructure/comms/src/connection/net_address/mod.rs

```diff
@@ -41,51 +59,90 @@ pub enum NetAddress {
     I2P(I2PAddress),
 }
 
+impl NetAddress {
+    /// Returns true if the [`NetAddress`] is an IP address, otherwise false
+    pub fn is_ip(&self) -> bool {
+        match *self {
+            NetAddress::IP(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if the [`NetAddress`] is a Tor Onion address, otherwise false
+    pub fn is_tor(&self) -> bool {
+        match *self {
+            NetAddress::Tor(_) => true,
+            _ => false,
+        }
+    }
+
+    /// Returns true if the [`NetAddress`] is an I2P address, otherwise false
+    pub fn is_i2p(&self) -> bool {
+        match *self {
+            NetAddress::I2P(_) => true,
+            _ => false,
+        }
+    }
+}
+
 impl FromStr for NetAddress {
     type Err = NetAddressError;
 
+    /// Parses a [`str`] into a [`NetAddress`]
     fn from_str(address: &str) -> Result<Self, Self::Err> {
         if let Ok(addr) = address.parse::<SocketAddress>() {
```

Comment:
How does rust handle errors?
I know in most other languages using a try catch actively in code can be very slow.

It might be worth investigating and perhaps doing something more like string matching than catching the failure

Reply:
In Rust it's as fast as returning an enum (there are no try catch jumps) - any slowness would be from running the parsers in sequence, but the parsers are really quick (no regexes). We can avoid using this method if we know the address type ahead of time and use the `From` implementations below, however I still think the feature of being able to parse any string into a `NetAddress` is useful.